src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -11,21 +11,526 @@
+	frelGraph->isValid = false;
+
+	/* create (oid) -> [FRelNode] hash */
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(Oid);
+	info.entrysize = sizeof(FRelNode);
+	info.hash = oid_hash;
+	info.hcxt = CurrentMemoryContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT);
+
+	frelGraph->nodeMap = hash_create("foreign key relation map (oid)",
+									 64 * 32, &info, hashFlags);
+
+
+	PopulateAdjacencyLists();
+
+	frelGraph->isValid = true;
+	MemoryContextSwitchTo(oldContext);
+}
+
+
+/*
+ * PopulateAdjacencyLists gets foreign key information from pg_constraint
+ * metadata table and populates them to the foreign key relation graph.
+ */
+static void
+PopulateAdjacencyLists(void)
+{
+	SysScanDesc scanDescriptor;
+	HeapTuple tuple;
+	Relation pgConstraint;
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 1;
+
+	Oid prevReferencingOid = InvalidOid;
+	Oid prevReferencedOid = InvalidOid;
+	List *frelEdgeList = NIL;
+	ListCell *frelEdgeCell = NULL;
+
+	pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
+
+	ScanKeyInit(&scanKey[0], Anum_pg_constraint_contype, BTEqualStrategyNumber, F_CHAREQ,
+				CharGetDatum(CONSTRAINT_FOREIGN));
+	scanDescriptor = systable_beginscan(pgConstraint, InvalidOid, false,
+										NULL, scanKeyCount, scanKey);
+
+	while (HeapTupleIsValid(tuple = systable_getnext(scanDescriptor)))
+	{
+		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(tuple);
+		FRelEdge *currentFRelEdge = NULL;
+
+		currentFRelEdge = palloc(sizeof(FRelEdge));
+		currentFRelEdge->referencingRelationOID = constraintForm->conrelid;
+		currentFRelEdge->referencedRelationOID = constraintForm->confrelid;
+
+		frelEdgeList = lappend(frelEdgeList, currentFRelEdge);
+	}
+
+	/*
+	 * Since there is no index on columns we are planning to sort tuples
+	 * sorting tuples manually instead of using scan keys
+	 */
+	frelEdgeList = SortList(frelEdgeList, CompareFRelEdges);
+
+	foreach(frelEdgeCell, frelEdgeList)
+	{
+		FRelEdge *currentFRelEdge = (FRelEdge *) lfirst(frelEdgeCell);
+
+		/* we just saw this edge, no need to add it twice */
+		if (currentFRelEdge->referencingRelationOID == prevReferencingOid &&
+			currentFRelEdge->referencedRelationOID == prevReferencedOid)
+		{
+			continue;
+		}
+
+		AddEdge(frelGraph->nodeMap, currentFRelEdge->referencingRelationOID,
+				currentFRelEdge->referencedRelationOID);
+
+		prevReferencingOid = currentFRelEdge->referencingRelationOID;
+		prevReferencedOid = currentFRelEdge->referencedRelationOID;
+	}
+
+	systable_endscan(scanDescriptor);
+	heap_close(pgConstraint, AccessShareLock);
+}
+
+
+/* Helper function to compare two FRelEdges using referencing and referenced ids
+ * respectively */
+static int
+CompareFRelEdges(const void *leftElement, const void *rightElement)
+{
+	const FRelEdge *leftEdge = *((const FRelEdge **) leftElement);
+	const FRelEdge *rightEdge = *((const FRelEdge **) rightElement);
+
+	int referencingDiff = leftEdge->referencingRelationOID -
```

Comment:
I'm almost sure that `int = unsigned - unsigned` works fine even if the result is negative.

But, could you please double check that?

(oid is unsigned int, the result of the operation is signed, so make sure that we don't have any subtle bugs because of it)

Reply:
Let's do not forget to double check this.